File: KafkaBasicsAndBeyond/TrafficMonitoringApp/src/main/java/com/github/leosilvadev/traffic_monitoring/Detector.java
package com.github.leosilvadev.traffic_monitoring;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

public class Detector {

    private final AtomicBoolean started;
    private final Random random;
    private final String sensorId;

    public Detector(final String sensorId) {
        this.started = new AtomicBoolean(false);
        this.random = new Random();
        this.sensorId = sensorId;
    }

    public void start(final Consumer<Detection> onDetection) {
        if (started.compareAndSet(false, true)) {
            System.out.println("Starting Detector...");
            new Thread(() -> {
                while (started.get()) {
                    onDetection.accept(detect());
                    waitFor(random.nextInt(100));
                }
            }).start();
        } else {
            System.out.println("Detector is already running");
        }
    }

    public void stop() {
        if (started.compareAndSet(true, false)) {
            System.out.println("Stopping Detector...");
        } else {
            System.out.println("Detector is already stopped");
        }
    }

    public Detection detect() {
        final var detectionId = sensorId + "_" + System.currentTimeMillis();
        return new Detection(
                sensorId,
                detectionId,
                "".getBytes(),
                random.nextDouble() * 150
        );
    }

    public String getSensorId() {
        return sensorId;
    }

    private static void waitFor(final Integer millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            //never mind...
        }
    }
}

File: KafkaBasicsAndBeyond/TrafficMonitoringAppNode/app/detector.js
class Detection {
    constructor(sensorId, id, image, speed) {
        this.sensorId = sensorId;
        this.id = id;
        this.image = image;
        this.speed = speed;
    }
}

class Detector {
    constructor(sensorId) {
        this.sensorId = sensorId;
        this.started = false;
    }

    detect() {
        const id = `${this.sensorId}_${Date.now()}`;
        const speed = Math.random(150) * 100;
        return new Detection(this.sensorId, id, Buffer.of(), speed);
    }

    start(onDetect) {
        if (this.started) {
            console.warn('Detector is already running');
            return;
        }

        this.started = true;
        this.run(onDetect);
    }

    run(onDetect) {
        if (this.started) {
            setTimeout(() => {
                onDetect(this.detect());
                this.run(onDetect);
            }, Math.random(100) * 1000);
        }
    }
}

module.exports = sensorId => new Detector(sensorId)